src/power_grid_model/data_types.py

```diff
@@ -17,64 +17,171 @@
+        - the scenario 0 sets the statuses of components with ids 0 and 1 to 1 (and keeps defaults for other components)
         - scenario 1 keeps the default values for all components
         - scenario 2 sets the statuses of component with id 0 to 0 (and keeps defaults for other components)
 """
 
+SparseBatchColumnarData = Dict[str, Union[IndexPointer, SingleColumnarData]]
+"""
+Sparse batch columnar data is a dictionary containing the keys `indptr` and `data`.
+
+- data: a :class:`SingleColumnarData`. The exact supported attribute columns depend on the component type.
+- indptr: an :class:`IndexPointer` representing the start and end indices for each batch scenario.
+
+- Examples:
+
+    - structure: {"indptr": :class:`IndexPointer`, "data": :class:`SingleArray`}
+    - concrete example: {"indptr": [0, 2, 2, 3], "data": {"id": [0, 1, 0], "status": [1, 1, 0]}}
+
+        - the scenario 0 sets the status of components with ids 0 and 1 to 1 (and keeps defaults for other components)
+        - scenario 1 keeps the default values for all components
+        - scenario 2 sets the status of component with id 0 to 0 (and keeps defaults for other components)
+"""
+
+DenseBatchData = Union[DenseBatchArray, DenseBatchColumnarData]
+"""
+Dense batch data can be a :class:`DenseBatchArray` or a :class:`DenseBatchColumnarData`.
+"""
+
+SparseBatchData = Union[SparseBatchArray, SparseBatchColumnarData]
+"""
+Sparse batch data can be a :class:`SparseBatchArray` or a :class:`SparseBatchColumnarData`.
+"""
+
 BatchArray = Union[DenseBatchArray, SparseBatchArray]
 """
 A batch array is a either a :class:`DenseBatchArray` or a :class:`SparseBatchArray`.
 """
 
+BatchColumnarData = Union[DenseBatchColumnarData, SparseBatchColumnarData]
+"""
+Batch columnar data is either a :class:`DenseBatchColumnarData` or a :class:`SparseBatchColumnarData`.
+"""
+
 DataArray = Union[SingleArray, BatchArray]
 """
 A data array can be a :class:`SingleArray` or a :class:`BatchArray`.
 """
 
-SingleDataset = Dict[ComponentType, SingleArray]
+ColumnarData = Union[SingleColumnarData, BatchColumnarData]
+"""
+Columnar data can be :class:`SingleColumnarData` or :class:`BatchColumnarData`.
+"""
+
+# SingleComponentData = Union[SingleArray, SingleColumnarData]
+# """
+# Single component data can be :class:`SingleArray` or :class:`SingleColumnarData`.
+# """
+SingleComponentData = Union[SingleArray]
+"""
+Single component data is a :class:`SingleArray`.
+"""
+
+# BatchComponentData = Union[BatchArray, BatchColumnarData]
+# """
+# Batch component data can be :class:`BatchArray` or :class:`BatchColumnarData`.
+# """
+BatchComponentData = Union[BatchArray]
+"""
+Batch component data is a :class:`BatchArray`.
+"""
+
+# ComponentData = Union[DataArray, ColumnarData]
+# """
+# Component data can be :class:`DataArray` or :class:`ColumnarData`.
+# """
+ComponentData = Union[DataArray]
+"""
+Component data is a :class:`DataArray`.
+"""
```

Comment:
this prevents using in existing functionality. to be added when python API is made ready for columnar data